File: b2b_project/backend/B2BApi/Models/Product.cs
using System.ComponentModel.DataAnnotations;

namespace B2BApi.Models
{
    public class Product
    {
        public int Id { get; set; }
        
        [Required]
        [StringLength(50)]
        public string ProductCode { get; set; } = string.Empty;
        
        [Required]
        [StringLength(500)]
        public string Name { get; set; } = string.Empty;
        
        // Fiyat Bilgileri
        public decimal ListPrice { get; set; } = 0; // Liste Fiyatı
        
        [Required]
        public decimal BuyPriceExcludingVat { get; set; } = 0; // KDV Hariç Alış Fiyatı
        
        [Required]
        public decimal BuyPriceIncludingVat { get; set; } = 0; // KDV Dahil Alış Fiyatı
        
        [Required]
        public decimal MyPrice { get; set; } = 0; // Kar marjı eklenmiş satış fiyatımız (KDV Dahil)

        // Hesaplanmış satış fiyatları (JSON'a dahil edilecek)
        public decimal SalePriceExcludingVat => VatRate > 0 ? MyPrice / (1 + VatRate / 100) : MyPrice;
        public decimal SalePriceIncludingVat => MyPrice;

        // İskonto Bilgileri
        public decimal Discount1 { get; set; } = 0;
        public decimal Discount2 { get; set; } = 0;
        public decimal Discount3 { get; set; } = 0;
        
        // KDV Bilgisi
        public decimal VatRate { get; set; } = 0; // KDV Oranı
        
        // Resim Bilgileri
        [StringLength(500)]
        public string? ImageUrl { get; set; } // Orijinal resim URL'i
        
        [StringLength(200)]
        public string? LocalImagePath { get; set; } // Lokal resim path'i
        
        [Required]
        public decimal MarginPercentage { get; set; } = 40;
        
        [Required]
        public DateTime LastUpdated { get; set; } = DateTime.UtcNow;

        // Soft Delete için
        public bool IsDeleted { get; set; } = false;

        public DateTime? DeletedAt { get; set; }

        // Backward compatibility için eski alanları kaldırdık
        // BuyPrice -> BuyPriceExcludingVat
        // Stock -> artık kullanmıyoruz
        // Category -> artık kullanmıyoruz
    }
}

File: b2b_project/frontend/lib/models/product.dart
class Product {
  final int id;
  final String productCode;
  final String name;
  final double listPrice;
  final double buyPriceExcludingVat;
  final double buyPriceIncludingVat;
  final double myPrice;
  final double discount1;
  final double discount2;
  final double discount3;
  final double vatRate;
  final String? imageUrl;
  final String? localImagePath;
  final double marginPercentage;
  final DateTime lastUpdated;
  final bool isDeleted;
  final DateTime? deletedAt;

  Product({
    required this.id,
    required this.productCode,
    required this.name,
    required this.listPrice,
    required this.buyPriceExcludingVat,
    required this.buyPriceIncludingVat,
    required this.myPrice,
    required this.discount1,
    required this.discount2,
    required this.discount3,
    required this.vatRate,
    this.imageUrl,
    this.localImagePath,
    required this.marginPercentage,
    required this.lastUpdated,
    this.isDeleted = false,
    this.deletedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as int,
      productCode: json['productCode'] as String,
      name: json['name'] as String,
      listPrice: _parseDouble(json['listPrice']),
      buyPriceExcludingVat: _parseDouble(json['buyPriceExcludingVat']),
      buyPriceIncludingVat: _parseDouble(json['buyPriceIncludingVat']),
      myPrice: _parseDouble(json['myPrice']),
      discount1: _parseDouble(json['discount1']),
      discount2: _parseDouble(json['discount2']),
      discount3: _parseDouble(json['discount3']),
      vatRate: _parseDouble(json['vatRate']),
      imageUrl: json['imageUrl'] as String?,
      localImagePath: json['localImagePath'] as String?,
      marginPercentage: _parseDouble(json['marginPercentage']),
      lastUpdated: DateTime.parse(json['lastUpdated'] as String),
      isDeleted: json['isDeleted'] as bool? ?? false,
      deletedAt: json['deletedAt'] != null ? DateTime.parse(json['deletedAt'] as String) : null,
    );
  }

  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) {
      return double.tryParse(value) ?? 0.0;
    }
    return 0.0;
  }

  double get salePriceExcludingVat => vatRate > 0 ? myPrice / (1 + vatRate / 100) : myPrice;
  double get salePriceIncludingVat => myPrice;
}
